packages/perseus-editor/src/widgets/free-response-editor.tsx

```diff
@@ -32,18 +39,148 @@ class FreeResponseEditor extends React.Component<Props> {
         return warnings;
     }
 
+    handleUpdateCriterion = (
+        index: number,
+        criterion: PerseusFreeResponseWidgetScoringCriterion,
+    ) => {
+        const newCriteria = this.props.scoringCriteria.map((c, i) => {
+            if (i === index) {
+                return criterion;
+            }
+            return c;
+        });
+
+        this.props.onChange({
+            scoringCriteria: newCriteria,
+        });
+    };
+
+    handleDeleteCriterion = (index: number) => {
+        this.props.onChange({
+            scoringCriteria: this.props.scoringCriteria.filter(
+                (_, i) => i !== index,
+            ),
+        });
+    };
+
+    handleAddCriterion = () => {
+        this.props.onChange({
+            scoringCriteria: [...this.props.scoringCriteria, {text: ""}],
+        });
+    };
+
+    renderCriteriaList = () => {
+        return this.props.scoringCriteria.map((criterion, index) => {
+            return (
+                <CriterionEditor
+                    criterion={criterion}
+                    index={index}
+                    key={index}
+                    numCriteria={this.props.scoringCriteria.length}
```

Comment:
Ultra tiny nit but I feel like telling CriterionEditor how many criteria there are is TMI. Really what you're saying is something along the lines of `deletable`.

```suggestion
                    deletable={this.props.scoringCriteria.length <= 1}
```

Go with your gut on this one; feel free to ignore.